rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromMockObjectRector.php

```diff
@@ -0,0 +1,153 @@
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Rector\PhpParser\Node\BetterNodeFinder;
+use Rector\Rector\AbstractRector;
+use Rector\ValueObject\PhpVersionFeature;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromMockObjectRector\ReturnTypeFromMockObjectRectorTest
+ */
+final class ReturnTypeFromMockObjectRector extends AbstractRector implements MinPhpVersionInterface
+{
+    /**
+     * @var string
+     */
+    private const MOCK_OBJECT_CLASS = 'PHPUnit\Framework\MockObject\MockObject';
+
+    public function __construct(
+        private readonly BetterNodeFinder $betterNodeFinder
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add known property and return MockObject types', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeTest extends TestCase
+{
+    public function test()
+    {
+        $someMock = $this->createMock(SomeClass::class);
+        return $someMock;
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+class SomeTest extends TestCase
+{
+    public function test(): \PHPUnit\Framework\MockObject\MockObject
+    {
+        $someMock = $this->createMock(SomeClass::class);
+        return $someMock;
+    }
+}
+CODE_SAMPLE
+            ),
+
+        ]);
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        // type is already known
+        if ($node->returnType instanceof Node) {
+            return null;
+        }
+
+        if (! $this->isInsideTestCaseClass($node)) {
+            return null;
+        }
+
+        // we need exactly 1 return
+        $returns = $this->betterNodeFinder->findReturnsScoped($node);
+        if (count($returns) !== 1) {
+            return null;
+        }
+
+        $soleReturn = $returns[0];
+        if (! $soleReturn->expr instanceof Expr) {
+            return null;
+        }
+
+        $returnType = $this->getType($soleReturn->expr);
+        if (! $this->isMockObjectType($returnType)) {
+            return null;
+        }
+
+        $node->returnType = new FullyQualified(self::MOCK_OBJECT_CLASS);
+
+        return $node;
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::SCALAR_TYPES;
+    }
+
+    private function isIntersectionWithMockObjectType(Type $type): bool
+    {
+        if (! $type instanceof IntersectionType) {
+            return false;
+        }
+
+        if (count($type->getTypes()) !== 2) {
+            return false;
+        }
+
+        return in_array(MockObject::class, $type->getObjectClassNames());
```

Comment:
use string constant to avoid prefixed on downgrade

```suggestion
        return in_array(self::MOCK_OBJECT_CLASS, $type->getObjectClassNames());
```